File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/HashUtils.swift
import CryptoKit
import Foundation

enum HashUtilsError: Error {
  case couldNotOpenFile
}

enum HashUtils {
  static func sha256(fileURL: URL) throws -> String {
    guard let handle = try? FileHandle(forReadingFrom: fileURL) else {
      throw HashUtilsError.couldNotOpenFile
    }
    defer { try? handle.close() }

    var hasher = SHA256()

    while autoreleasepool(invoking: {
      let data = handle.readData(ofLength: 1024 * 1024)
      if data.isEmpty {
        return false
      }
      hasher.update(data: data)
      return true
    }) {}

    return hasher.finalize().map { String(format: "%02x", $0) }.joined()
  }

  static func verify(fileURL: URL, expectedSha256: String) throws -> Bool {
    let actual = try sha256(fileURL: fileURL)
    return actual.lowercased() == expectedSha256.lowercased()
  }

  static func verifyDownload(fileURL: URL, expectedSha256: String, expectedBytes: Int?, kind: String) throws {
    let actual = try sha256(fileURL: fileURL)
    let attributes = try FileManager.default.attributesOfItem(atPath: fileURL.path)
    guard let size = attributes[.size] as? NSNumber else { throw HashUtilsError.couldNotOpenFile }
    let received = size.int64Value
    let hashMatches = actual.caseInsensitiveCompare(expectedSha256) == .orderedSame
    let sizeMatches = expectedBytes.map { Int64($0) == received } ?? true
    guard hashMatches && sizeMatches else {
      let safeExpected = expectedSha256.range(of: "^[a-fA-F0-9]{64}$", options: .regularExpression) != nil
        ? expectedSha256.lowercased() : "invalid"
      let reason = hashMatches ? "size mismatch" : "hash mismatch"
      throw NSError(domain: "OtaKit", code: 1, userInfo: [NSLocalizedDescriptionKey:
        "Downloaded \(kind) \(reason); expectedSha256=\(safeExpected); actualSha256=\(actual); expectedBytes=\(expectedBytes ?? -1); receivedBytes=\(received)"])
    }
  }
}

File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/HashUtils.java
package com.otakit.updater;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.security.MessageDigest;

final class HashUtils {

  private HashUtils() {}

  static String sha256(InputStream input) throws Exception {
    MessageDigest digest = MessageDigest.getInstance("SHA-256");
    byte[] buffer = new byte[1024 * 1024];
    int read;
    while ((read = input.read(buffer)) > 0) {
      digest.update(buffer, 0, read);
    }
    byte[] hash = digest.digest();
    StringBuilder builder = new StringBuilder();
    for (byte b : hash) {
      builder.append(String.format("%02x", b));
    }
    return builder.toString();
  }

  static String sha256(File file) throws Exception {
    MessageDigest digest = MessageDigest.getInstance("SHA-256");
    try (FileInputStream input = new FileInputStream(file)) {
      byte[] buffer = new byte[1024 * 1024];
      int read;
      while ((read = input.read(buffer)) > 0) {
        digest.update(buffer, 0, read);
      }
    }

    byte[] hash = digest.digest();
    StringBuilder builder = new StringBuilder();
    for (byte b : hash) {
      builder.append(String.format("%02x", b));
    }
    return builder.toString();
  }

  static boolean verify(File file, String expectedSha256) throws Exception {
    return sha256(file).equalsIgnoreCase(expectedSha256);
  }

  static void verifyDownload(File file, String expectedSha256, long expectedBytes, String kind)
    throws Exception {
    String actual = sha256(file);
    boolean hashMatches = actual.equalsIgnoreCase(expectedSha256);
    boolean sizeMatches = expectedBytes < 0 || expectedBytes == file.length();
    if (!hashMatches || !sizeMatches) {
      String safeExpected =
        expectedSha256 != null && expectedSha256.matches("[a-fA-F0-9]{64}")
          ? expectedSha256.toLowerCase(java.util.Locale.ROOT)
          : "invalid";
      throw new IllegalStateException(
        "Downloaded " +
          kind +
          (hashMatches ? " size mismatch" : " hash mismatch") +
          "; expectedSha256=" +
          safeExpected +
          "; actualSha256=" +
          actual +
          "; expectedBytes=" +
          expectedBytes +
          "; receivedBytes=" +
          file.length()
      );
    }
  }
}
